services/mac_manager/src/main/java/com/futurewei/alcor/macmanager/service/implement/MacServiceImpl.java

```diff
@@ -62,236 +65,337 @@
+        if (macState == null)
+            throw (new ParameterNullOrEmptyException(MacManagerConstant.MAC_EXCEPTION_PARAMETER_NULL_EMPTY));
+        String rangeId = MacManagerConstant.DEFAULT_RANGE;
+        try {
+            macState = createMacStateInRange(rangeId, macState);
+        } catch (CacheException e) {
+            throw new MacRepositoryTransactionErrorException(MacManagerConstant.MAC_EXCEPTION_REPOSITORY_EXCEPTION);
+        } catch (MacRangeInvalidException | MacAddressUniquenessViolationException | MacAddressFullException | MacAddressRetryLimitExceedException e) {
             throw e;
         }
         return macState;
     }
 
-    public MacState createMacState(MacState macState) throws Exception {
+    /**
+     * create a MacState
+     *
+     * @param macState MAC allocation state
+     * @param rangeId  MAC address range to create
+     * @return MAC allocation state with an allocated MAC address
+     * @throws ParameterNullOrEmptyException          parameter macAddress is null or empty
+     * @throws MacRepositoryTransactionErrorException error during repository transaction
+     * @throws MacRangeInvalidException               mac range to create a new MAC address is null or not existing
+     * @throws MacAddressUniquenessViolationException MAC address is not unique. MAC address should be allocated to only one port
+     * @throws MacAddressFullException                All MAC addresses are created.
+     * @throws MacAddressRetryLimitExceedException    MAC addresss creation is tried more than limit
+     */
+    @Override
+    public MacState createMacStateInRange(String rangeId, MacState macState) throws ParameterNullOrEmptyException, MacRepositoryTransactionErrorException, MacRangeInvalidException, MacAddressUniquenessViolationException, MacAddressRetryLimitExceedException, MacAddressFullException {
         if (macState == null)
-            throw (new ParameterNullOrEmptyException(MacUtil.MAC_EXCEPTION_PARAMETER_NULL_EMPTY));
+            throw (new ParameterNullOrEmptyException(MacManagerConstant.MAC_EXCEPTION_PARAMETER_NULL_EMPTY));
         MacAddress macAddress = new MacAddress();
         if (macState.getState() == null)
-            macState.setState(MacUtil.MAC_STATE_ACTIVE);
+            macState.setState(MacManagerConstant.MAC_STATE_ACTIVE);
         else if (macState.getState().trim().length() == 0)
-            macState.setState(MacUtil.MAC_STATE_ACTIVE);
-        if (macPoolRepository.getSize() < (nMacPoolSize - 10)) {
-            CompletableFuture<Long> completableFuture = CompletableFuture.supplyAsync(() -> {
-                long n = 0;
-                try {
-                    n = generateMacInPool(20);
-                } catch (Exception e) {
+            macState.setState(MacManagerConstant.MAC_STATE_ACTIVE);
 
+        try {
+            MacRange range = macRangeRepository.findItem(rangeId);
+            if (range == null) {
```

Comment:
Please add brief comments here to describe the algorithm.

Reply:
Done
            //MAC address is generated in a MAC range betwen from ~ to, if there is no active user defined MAC range, by default default MAC range is applied.
            //Default MAC range: name is defined MacManagerConstant class, range is entire oui range. e.g.)
            // if oui=AA-BB-CC then default MAC range is AA-BB-CC-00-00-00 ~ AA-BB-CC-FF-FF-FF.

Reply:
Thanks. You could add those comments along with the codes, instead of in the PR.

Reply:
I added the comment to the code, too. This is just for your comfort.

Reply:
See it now. Thanks.